R/missing_package_linter.R

```diff
@@ -0,0 +1,44 @@
+#' @describeIn linters checks for missing packages in \code{library()},
+#'   \code{require()}, \code{loadNamespace()} and \code{requireNamespace()} calls.
+#' @export
+missing_package_linter <- function(source_file) {
+
+  if (!length(source_file$parsed_content)) return(list())
+
+  xml <- source_file$xml_parsed_content
+
+  library_calls <- c("library", "require", "loadNamespace", "requireNamespace")
+
+  name_xpath <- "OP-LEFT-PAREN[1]/following-sibling::expr[1][SYMBOL | STR_CONST]"
+  call_xpath <- paste0(sprintf(
+    "//expr[expr[SYMBOL_FUNCTION_CALL[%s]]/following-sibling::%s]",
+    paste0(sprintf("text()='%s'", library_calls), collapse = " or "),
+    name_xpath
+  ))
+
+  pkg_calls <- xml2::xml_find_all(xml, call_xpath)
+  pkg_names <- xml2::xml_find_all(pkg_calls, name_xpath)
+  pkg_names <- xml2::xml_text(pkg_names)
+  pkg_names <- parse(text = pkg_names, keep.source = FALSE)
+  pkg_names <- vapply(pkg_names, as.character, character(1L))
+
+  installed_packges <- .packages(all.available = TRUE)
```

Comment:
could this lib.paths be an argument to this linter? or perhaps known_packages as an argument. would make it more flexible to check code statically in scenarios where the packages may not be installed or the linter is run with a funny PATH

Reply:
https://github.com/jimhester/lintr/pull/551#discussion_r517034082